File: framework/kotlin/src/main/kotlin/soda/kotlin/web/WorkHandler.kt
package soda.kotlin.web

import com.sun.net.httpserver.HttpExchange
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import kotlin.reflect.full.memberFunctions
import kotlin.reflect.full.primaryConstructor

class WorkHandler : BaseHandler() {

    private val maxTimeoutMillis: Long = 9000

    private val timeoutMillis = maxTimeoutMillis

    override fun handleWork(exchange: HttpExchange): String {
        val content = readPostBody(exchange)
        Logger.info("request: $content")

        val req = Json.decodeFromString<WorkRequest>(content)
        val task = fun(): String {
            val classLoader = ClassLoaderCache.getForJar(req.classpath)
                ?: throw RuntimeException("no class loader for ${req.classpath}")
            val className = req.bootClass
            val kClass = classLoader.loadClass(className).kotlin
            val ctor = kClass.primaryConstructor
                ?: throw NoSuchMethodError("no constructor for class $className")
            val obj = ctor.call()
            val invoke = kClass.memberFunctions.find { it.name == "invoke" }
                ?: throw NoSuchMethodError("no method invoke found for class $className")
            return invoke.call(obj, req.testCase) as String
        }

        val job = TimeLimitedWork(task)
        val future = job.start()
        try {
            future.get(timeoutMillis, TimeUnit.MILLISECONDS)
        } catch (ex: TimeoutException) {
            job.kill()
            throw RuntimeException("work execution timeout", ex)
        }

        return ""
    }
}

File: framework/scala/src/main/scala/soda/scala/web/WorkHandler.scala
package soda.scala.web

import com.sun.net.httpserver.HttpExchange
import java.util.concurrent.{TimeoutException, TimeUnit}
import scala.reflect.runtime.universe._

import play.api.libs.json._

class WorkHandler extends BaseHandler {

  val maxTimeoutMillis: Long = 10000

  private val timeoutMillis: Long = maxTimeoutMillis

  override def handleWork(exchange: HttpExchange): String = {
    val content = readPostBody(exchange)
    Logger.info(s"request: $content")
    implicit val _format: OFormat[WorkRequest] = Json.format[WorkRequest]
    val jr = Json.parse(content).as[WorkRequest]

    val task: () => String = () => {
      val oldCtxLoader = Thread.currentThread().getContextClassLoader
      try {
        val classLoader = ClassLoaderCache.getForJar(jr.classpath)
        Thread.currentThread().setContextClassLoader(classLoader)
        val rm = runtimeMirror(classLoader)
        val csym = rm.staticClass(jr.bootClass)
        val classType = csym.typeSignature
        val cm = rm.reflectClass(csym)
        val ctor = cm.reflectConstructor(classType.decl(termNames.CONSTRUCTOR).asMethod)
        val obj = ctor.apply()
        val im = rm.reflect(obj)
        val applySym = classType.decl(TermName("apply"))
        val mm = im.reflectMethod(applySym.asMethod)
        mm.apply(jr.testCase).asInstanceOf[String]
      } finally {
        Thread.currentThread().setContextClassLoader(oldCtxLoader)
      }
    }

    val job = new TimeLimitedWork(task)
    val future = job.start()
    try {
      future.get(timeoutMillis, TimeUnit.MILLISECONDS)
    } catch {
      case ex: TimeoutException =>
        job.kill()
        throw new RuntimeException("Work execution timeout", ex)
    }
  }
}
